### PR TITLE
MG-2053 - Structure API Endpoints Based on Entities

### LINKED ISSUE #2053: Feature: Enhance OpenAPI Specification by Structuring Endpoints Based on Entities for Improved Clarity and User Experience
### Is your feature request related to a problem? Please describe.

Structuring endpoints based on entities rather than service-specific endpoints, the proposal seeks to streamline the understanding of functionalities related to specific entities, such as users, things, groups, channels, domains making it easier for users to navigate and comprehend the available API functionalities within a specific context.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Structuring API endpoints not based on service-specific endpoints but rather on entities, For example, within the `users.yml` file, all endpoints would start with /users. Operations related to specific users, such as retrieving user tags (`users/{userID}/tag`), would be included here. However, I propose including additional endpoints, like `/users/{userID}/things`, which pertain to the `things` associated with a user, even if the `things` functionality resides in a different service.

By including such endpoints in the `users.yml` API specification, users interacting with this document will find it easier to understand as they would expect `users.yml` to provide comprehensive user-related functionalities, even if certain functionalities are fulfilled by other services.

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by felixgateru**
# What type of PR is this?
This is a refactor as it restructures endpoints based on entities to improve clarity and user experience.

## What does this do?
This pr removes the following endpoints:
- /users/{memberID}/domains
- /channels/{chanID}/things
- /things/{thingID}/channels
- /users/{memberID}/channels
- /groups/{memberID}/channels
- /groups/{groupID}/users
- /channels/{channelID}/users
- /channels/{memberID}/groups
- /users/{memberID}/groups
- /domains/{domainID}/users

and replaces the functionlaity with list by entity option in list `domains`, `users` ,`groups`, `channels` and `things` as documented in the api.

## Which issue(s) does this PR fix/relate to?

- Resolves #2053 

## Have you included tests for your changes?
Yes, relevant tests have been added and updated.

## Did you document any new/modified feature?
Yes, the api documentation has been updated.

### CODE REVIEW

File: things/service.go [hunk 1]
```diff
@@ -172,17 +173,26 @@ func (svc service) ListClients(ctx context.Context, token, reqUserID string, pm
...
 		}
+		pm.IDs = thingIDs
 	}
 
-	if len(ids) == 0 && pm.Domain == "" {
+	if len(pm.IDs) == 0 && pm.Domain == "" {
 		return mgclients.ClientsPage{}, nil
 	}
-	pm.IDs = ids
 	tp, err := svc.clients.SearchClients(ctx, pm)
```
Comment (arvindh123) [inline things/service.go]: @felixgateru , Why we are using `SearchClients` instead of `RetrieveAll`

**Review by arvindh123 — CHANGES_REQUESTED** [review]
@felixgateru , I think list things is not in good shape, Could you for Domain member can able to list the things which they created ?

File: things/service.go (hunk 1 above)
Comment (felixgateru) [reply things/service.go]: This had been preffered in the latest implementation of list things. But I am checking which will be best for this.

File: things/service.go [hunk 2]
```diff
@@ -140,50 +141,68 @@ func (svc service) ViewClientPerms(ctx context.Context, token, id string) ([]str
...
-		}
-	default:
-		err := svc.checkSuperAdmin(ctx, res.GetUserId())
+	switch pm.EntityType {
+	case auth.UserType:
+		pm.EntityID = auth.EncodeDomainUserID(res.GetDomainId(), pm.EntityID)
+		return svc.listEntityThings(ctx, res.GetId(), res.GetDomainId(), auth.ViewPermission, auth.DomainType, res.GetDomainId(), pm)
+	case auth.GroupType:
+		pm.Permission = auth.GroupRelation
+		return svc.listEntityThings(ctx, res.GetId(), res.GetDomainId(), auth.ViewPermission, auth.GroupType, pm.EntityID, pm)
```
Comment (arvindh123) [inline things/service.go]: We can have authorization out of this function i.e before calling the function we can do authroization .

File: things/service.go [hunk 3]
```diff
@@ -140,50 +141,68 @@ func (svc service) ViewClientPerms(ctx context.Context, token, id string) ([]str
...
-			if _, err := svc.authorize(ctx, "", auth.UserType, auth.UsersKind, res.GetId(), auth.MembershipPermission, auth.DomainType, res.GetDomainId()); err != nil {
-				return mgclients.ClientsPage{}, err
-			}
-			ids, err = svc.listClientIDs(ctx, res.GetId(), pm.Permission)
-			if err != nil {
-				return mgclients.ClientsPage{}, errors.Wrap(svcerr.ErrNotFound, err)
-			}
+			pm.EntityType = auth.UserType
+			pm.EntityID = res.GetId()
+			return svc.listEntityThings(ctx, res.GetId(), res.GetDomainId(), auth.MembershipPermission, auth.DomainType, res.GetDomainId(), pm)
```
Comment (arvindh123) [inline things/service.go]: If domain is disabled , then this should not work for domain non administrator users. Please check this case

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: things/service.go (hunk 3 above)
Comment (felixgateru) [reply things/service.go]: I have checked this condition. When the domain is disabled only admins are able to list things

**Comment by dborovcanin** [comment]
I'm closing this one, since it will be implicitly resolved by https://github.com/absmach/magistrala/pull/2494.

**Pull request closed without merging**